[TLS / hw acceleration] AES ECB for NUCLEO_F439ZI

adustm opened the pull request:
## Description

This pull request contains the very first HW acceleration for STM devices
Could you review it and let us know if it fits the requested process for HW acceleration in mbedtls feature ?
I've also added aes in TESTS/mbedtls/selftest/main.cpp 
+---------------------------+--------+--------+--------+
| test case                 | passed | failed | result |
+---------------------------+--------+--------+--------+
| mbedtls_aes_self_test     | 1      | 0      | OK     |
| mbedtls_entropy_self_test | 1      | 0      | OK     |
| mbedtls_sha256_self_test  | 1      | 0      | OK     |
| mbedtls_sha512_self_test  | 1      | 0      | OK     |
+---------------------------+--------+--------+--------+

Once we agree on this files format and structure, we will be able to add soon AES-CBC mode, SHA1 etc...

## Status
**READY**

## Steps to test or reproduce
mbed test -m NUCLEO_F439ZI -t IAR -n tests-mbedtls-selftest

cc @bcostm @screamerbg @sbutcher-arm @0xc0170

jeromecoutant commented:
Hi
Is it OK to add some TARGET_xxx files in the features directory ?
Could we move these files into targets directory ?

adustm commented:
Hi @jeromecoutant 
It would be nice but I don't know the impact on code size etc...

@0xc0170 ?
Jerome and I would like to move mbedtls/target_STM code to the usual mbed-os/targets/... TARGET_STM directory.
Do you know if we can do that or not ? What is the impact of doing this ? 

Kind regards
Armelle

0xc0170 commented:
> Jerome and I would like to move mbedtls/target_STM code to the usual mbed-os/targets/... TARGET_STM directory.
Do you know if we can do that or not ? What is the impact of doing this ?

@sbutcher-arm Please review this PR.

screamerbg commented:
CC @RonEld

adustm commented:
Hello @RonEld @sbutcher-arm 
Could one of you review my PR please? 
Could you also let me know if we must keep STM's files in the feature directory or if we can move them to the targets usual STM directory ?

Cheers
Armelle

adustm commented:
Hi @sbutcher-arm 
Thank you for your comment, I'll wait for the review results.

> to separate out the code that uses the mbed HAL from the code that uses the mbed TLS interfaces

 features/mbedtls/targets/TARGET_STM/aes_alt.c is calling functions from targets/.../TARGET_STM32F4/device/stm32f4_hal_cryp.c 

so from ST's point of view, ST code is separated out

> Does this present any issues for you?

Let me try to explain the reason for this request from my colleagues and I. **If we can do it, let's go, if we can't, let it go. It is ARM to decide at the end** :-)

ST code is spread out in several features (mbedTLS / usb / lwip... ). 
As you may know we are providing support for more than 35 platforms. 
There are people working in parallel (from ST or not), not knowing into deep details the work of others. 

ST code is specific for our targets:
- we are handling the pins and call to hw functionalities for I2C / SPI / UART / etc... in targets directory
- We are handling the pins  and ST HW for the Ethernet phy in lwip feature directory + it calls ST HAL functions from targets directory
- We are handling the call to hw acceleration in mbedtls feature directory + it calls ST HAL functions from targets directory
- we are handling USB call in unsupported/USB directories
- And eventually we are handling the call to **hw entropy** in **targets** directory

When every code is located in a single directory tree we maximize the chance that every ST contributor is aware of the changes coming from others, and is better aware of the impact of his/her changes.

Anyway I have written this pull request following the specification (ST code in features/mbedtls).
The whole discussion is coming from an internal remark we had between ST people. We can now decide to take it into account or classify the discussion.

Kind regards
Armelle

sbutcher-arm commented:
Hi @adustm,

I think this is a useful discussion, and we have been seeking feedback from partners on the hardware acceleration interfaces. However, this is going outside the scope of this PR, and it may be more useful to discuss this directly.

To clarify - the targets directory is concerned with the mbed HAL and the interfaces mbed OS presents. mbed TLS has a life outside of mbed OS and is used on other platforms, and has separate and distinct interfaces itself. For instance, note mbed OS has a TRNG HAL API, whereas mbed TLS has it's own entropy interface, a mbed OS platform code brings the two interfaces together.

I think it's understandable that platform/SoC/driver code that might reside in the TARGETS directory may be called by driver code in the mbed TLS code driving the hardware acceleration code. I don't think that is an issue, and you are correct to say we could place the sources in other places (and indeed could move them later). This is really about logical organisation and management.

andresag01 commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,495 @@
+{
+
+    if ((ctx->nr != 10) && (ctx->nr != 12) && (ctx->nr != 14))
+        return(MBEDTLS_ERR_AES_INVALID_KEY_LENGTH);
```
Is this check necessary here? It seems that the functions that set the encryption and decryption key already do this checks. Also, please note that this is checking whether the number of rounds is valid and not the key length. Therefor, if there was a failure, the return code does not relate to the actual problem found.

adustm replied:
ok

andresag01 commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,495 @@
+        return(MBEDTLS_ERR_AES_INVALID_KEY_LENGTH);
+
+    /*------------------ AES Decryption ------------------*/
+    if(mode == MBEDTLS_AES_DECRYPT) /* AES decryption */
```
This comment is a little redundant...

adustm replied:
ok

andresag01 commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,495 @@
+        mbedtls_aes_decrypt( ctx, input, output );
+    }
+    /*------------------ AES Encryption ------------------*/
+    else /* AES encryption */
```
This comment is a little redundant

andresag01 commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,495 @@
+        // unaccelerated mode
+        //
+    }
+#endif
```
I think this block of code is only for x86 machines. Please consider removing it...

adustm replied:
ok

andresag01 commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,495 @@
+                          unsigned char output[16] )
+{
+    
+    HAL_CRYP_AESECB_Encrypt(&hcryp_aes, (uint8_t *)input, 16, (uint8_t *)output, 10);
```
I think it would be good to check this function's return code for errors.

adustm replied:
it does return timeout for instance. but mbedtls_aes_encrypt returns void. How do you usually handle it ?

sbutcher-arm replied:
I think we have an issue here. The interface was originally designed for a software implementation of AES and was later re-purposed for other implementations.

Here, I don't think the interface suits what we're trying to achieve in hardware, therefore I think we should deprecate this interface, and implement a new interface which enables a error return code. 

@andresag01 - Can you raise an issue/enhancement on the mbed TLS github repo?

@adustm - I think we can accept this code as it is at the moment, but we'll probably want to refine this code to work with a new interface later. In the meantime, can we remove the `mbedtls_printf()` and replace it with a comment?

andresag01 replied:
@sbutcher-arm: I have raised the new issue in the mbed TLS repository at https://github.com/ARMmbed/mbedtls/issues/817

andresag01 commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,495 @@
+                          unsigned char output[16] )
+{
+
+    HAL_CRYP_AESECB_Decrypt(&hcryp_aes, (uint8_t *)input, 16, (uint8_t *)output, 10);
```
I think it would be good to check this function's return code for errors.

adustm replied:
same comment

andresag01 commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,495 @@
+ */
+#if defined(MBEDTLS_AES_ALT)
+
+#include <stdio.h>
```
The code in this file uses a few macros such as `MBEDTLS_CIPHER_MODE_CTR` which are defined in `mbedtls/config.h`, but the file is not included here.

adustm replied:
thanks I will check these

screamerbg commented:
@adustm @sbutcher-arm I think there are 2 points of views:
* how features in mbed OS are organized
* how partner ports for targets in mbed OS organized

I agree with @adustm about having all target support code in one place under `targets`. The ST team is supporting significant amount of targets and having this separation is inconvenient, makes it harder for the mbed Partner teams to maintain (it's all over the place already) and reduces the chance for community contributions.

@sbutcher-arm wrote:
`mbed TLS has a life outside of mbed OS and is used on other platforms, and has separate and distinct interfaces itself.`

This PR is not to the mbed TLS specifically (e.g. this is not http://github.com/ARMmbed/mbedtls), but to extend an existing target support in mbed OS for TLS features. If the TLS team would like to request from ST to contribute this HW crypto support http://github.com/ARMmbed/mbedtls then please follow the formal process via email. 

Also if there is visibility issue, then this could be addressed by adding a label for the ARM mbed TLS team to review.

Perhaps this is a wider question (hence cc @sg-):
* where do features live - currently in /features
* where do drivers code live - currently in /targets for some, but /features for other
* where do glue layers live - currently /targets for HAL, LWIP /features/FEATURE_LWIP

Also it's important to understand the maintainability point of view, e.g. how much code mbed Partner teams are managing.

For example, the following structure may satisfy both ST team code maintenance and glue layers for features support

```
/targets
 -> TARGET_<vendor>
 ---> TARGET_<mcu_family>
-------> FEATURE_LWIP
-----------> all LWIP driver code for that MCU family
```

sg- commented:
Top level target code is organized to contain MCU SDKs and a binding to the mbed HAL. Any feature integrated as a mbed HAL interface and enabled by ```device_has``` should remain in ```feature/xxx/TARGET_xxx``` At the point a mbed HAL interface and tests come available for a features integration the implementation should move into the top level ```targets``` directory.

In this case, I agree with @sbutcher-arm 
> The idea of keeping the mbed TLS acceleration driver code in the features/mbedtls/targets directory is to separate out the code that uses the mbed HAL from the code that uses the mbed TLS interfaces, as well as making it clearer that code that is pushed to the HAL targets directory has a different approvals process from that pushed to the mbed TLS targets directory, as the cryptographic code needs an additional security review.

From the developer guide for hardware acceleration: https://docs.mbed.com/docs/mbed-os-handbook/en/latest/advanced/tls_hardware_acceleration/#how-to-implement-the-functions 
> Clone the mbed OS repository and copy the source code of your function definitions to the features/mbedtls/targets/TARGET_XXXX directory specific to your target. Create a pull request when your code is finished and production ready. You may create a directory structure similar to the one you have for the HAL if you feel it appropriate.

adustm commented:
Hello,
Thank you @screamerbg and @sg- for your answers and comments.
I think the discussion can be closed.

I'll start working on the PR review comments.
Kind regards

adustm commented on features/mbedtls/targets/TARGET_STM/TARGET_STM32F4/TARGET_NUCLEO_F439ZI/mbedtls_device.h:
```diff
@@ -0,0 +1,30 @@
+#define MBEDTLS_AES_SETKEY_ENC_ALT
+#define MBEDTLS_AES_SETKEY_DEC_ALT
+#define MBEDTLS_AES_ENCRYPT_ALT
+#define MBEDTLS_AES_DECRYPT_ALT
```
Ok I will try that

adustm commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,495 @@
+#include <stdio.h>
+#include "cmsis.h"
+#include "string.h"
+#include "aes.h"
```
ok

adustm commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,495 @@
+#include "cmsis.h"
+#include "string.h"
+#include "aes.h"
+CRYP_HandleTypeDef hcryp_aes;
```
you are totally right about the thread safety. I already have an issue with the CTR-DRBG self-test that is calling 2 instances of aes. I will rework it and declare CRY_HandleTypedef as part of the mbedtls_context_aes and remove the unnecessary fields (nr ...).
No more global (or even static )

adustm commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,495 @@
+ * \brief          Initialize AES context
+ *
+ * \param ctx      AES context to be initialized
+ */
```
ok

adustm commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,495 @@
+ *
+ * \return         0 if successful, or MBEDTLS_ERR_AES_INVALID_KEY_LENGTH
+ */
+#if defined(MBEDTLS_AES_SETKEY_ENC_ALT)
```
ok, the difference between full module alt and single function alt was not clear to me

adustm commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,495 @@
+    }
+
+    /* Deinitializes the CRYP peripheral */
+    HAL_CRYP_DeInit(&hcryp_aes);
```
every HAL_CRYP functions are returning a status (even TIMEOUT for the crpypto function itself), but most mbed functions don't have a return type. How would you like us to treat a TIMEOUT error for instance ?

adustm replied:
Could I create other MBEDTLS errors ?

adustm commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,495 @@
+ *
+ * \return         0 if successful, or MBEDTLS_ERR_AES_INVALID_KEY_LENGTH
+ */
+#if defined(MBEDTLS_AES_SETKEY_DEC_ALT)
```
ok

adustm commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,495 @@
+ */
+#if defined(MBEDTLS_AES_SETKEY_DEC_ALT)
+int mbedtls_aes_setkey_dec( mbedtls_aes_context *ctx, const unsigned char *key,
+                            unsigned int keybits )
```
ok

adustm commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,495 @@
+        break;
+      case 256:
+        ctx->nr = 14;
+        memcpy(ctx->aes_dec_key, key, 32);
```
you're right

adustm commented on features/mbedtls/targets/TARGET_STM/aes_alt.h:
```diff
@@ -0,0 +1,264 @@
+{
+    int nr;                        /*!<  number of rounds  */
+    uint32_t *rk;                  /*!<  AES round keys    */
+    uint32_t buf[68];              /*!<  unaligned data    */
```
rework on going to allow multi instance

adustm commented on features/mbedtls/targets/TARGET_STM/aes_alt.h:
```diff
@@ -0,0 +1,264 @@
+    unsigned char aes_enc_key[32]; /* Encryption key */
+    unsigned char aes_dec_key[32]; /* Decryption key */
+}
+mbedtls_aes_context;
```
ok

adustm commented on targets/targets.json:
```diff
@@ -899,7 +899,7 @@
         "supported_toolchains": ["ARM", "uARM", "GCC_ARM", "IAR"],
         "progen": {"target": "nucleo-f439zi"},
-        "macros": ["TRANSACTION_QUEUE_SIZE_SPI=2"],
+        "macros": ["TRANSACTION_QUEUE_SIZE_SPI=2", "MBEDTLS_CONFIG_HW_SUPPORT", "MBEDTLS_AES_ALT"],
```
I will try that

adustm replied:
Thank you very much @andresag01 for this detailled and interesting review :+1
I will rework that and send updates

adustm commented on features/mbedtls/targets/TARGET_STM/aes_alt.h:
```diff
@@ -0,0 +1,264 @@
+    uint32_t *rk;                  /*!<  AES round keys    */
+    uint32_t buf[68];              /*!<  unaligned data    */
+    unsigned char aes_enc_key[32]; /* Encryption key */
+    unsigned char aes_dec_key[32]; /* Decryption key */
```
you're right.

adustm commented:
Hi @andresag01 
I've reworked the code following your recommendations.
MBEDTLSxxx_ALT tags should be well handled
I've added AES_CBC hw acceleration
AND : I've fixed the 'not thread safe' issue you pointed out (and now I have CTR-DRBG ok)

![image](https://cloud.githubusercontent.com/assets/14051363/23074998/1f12473c-f53b-11e6-8da3-297bbbd29479.png)

I hope everything will be fine now. 
Kind regards
Armelle

0xc0170 commented:
Another round of review please?

retest uvisor

adustm commented:
Hello, is there anything blocking ?

0xc0170 commented:
@andresag01 @sbutcher-arm

RonEld commented:
I have looked in the changes, and I saw that there is many code duplicity, and copy\paste from the sw implementation. I beleive this is due to the current implementation of the HW configuration. This code duplication could have been resolved by `MBEDTLS_AES_SETKEY_ENC_ALT` , `MBEDTLS_AES_SETKEY_DEC_ALT` , `MBEDTLS_AES_ENCRYPT_ALT` , `MBEDTLS_AES_DECRYPT_ALT` , but the `mbedtls_aes_context` needs to be modified to support the HW AES, and this is currently not possible without `MBEDTLS_AES_ALT`.  
According to that, I opened a [github enhancement request](https://github.com/ARMmbed/mbedtls/issues/818) , that I believe could remove this duplicity.  
Other than that, the changes look good to me

sbutcher-arm commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,293 @@
+                          unsigned char output[16] )
+{
+    
+    if (HAL_CRYP_AESECB_Encrypt(&ctx->hcryp_aes, (uint8_t *)input, 16, (uint8_t *)output, 10) !=0)
```
We've identified that we need to change this function so it can return an error code, but until we do, I don't think we should be calling `mbedtls_printf()` here, unless it's restricted to be in just debug builds or similar.

adustm replied:
hello @sbutcher-arm 
Do you suggest that I remove the mbedtls_printf and replace it by a comment line 
`/* return error code once mbedtls_aes_encrypt and mbedtls_aes_decrypt functions will return one */`
?

sbutcher-arm commented on features/mbedtls/targets/TARGET_STM/aes_alt.c:
```diff
@@ -0,0 +1,293 @@
+{
+
+    if(HAL_CRYP_AESECB_Decrypt(&ctx->hcryp_aes, (uint8_t *)input, 16, (uint8_t *)output, 10))
+        mbedtls_printf( "HAL_CRYP_AESECB_Decrypt timeout\n" );
```
Same point as above on `mbedtls_printf()`.

adustm replied:
same question as above

sbutcher-arm commented:
Hi @adustm,

Looking over the code, and based on the feedback from @andresag01 and @RonEld, I can see we need to refine the hardware acceleration interface that mbed TLS presents. I don't want those changes to hold up this work, so we'll do that independently and then later can revise this code to take advantage of those changes. Please feel free to provide more feedback on those changes.

Thanks for your patience as we've worked through this pull request.

My only concern is the calls to `mbedtls_printf()` in case of an error, but that more shows a problem in our interfaces than your code.

adustm commented:
Hello @RonEld 
Your comment about thread safety is not clear to me and the PR you're pointing at is not merged.
Do you have a document, or a code example that I could read to understand better ?

Thank you to the 3 reviewers
Kind regards

adustm commented:
Hello, I've removed the mbedtls_printf error notifications
I did not change the thread safety nor the code dupplicity as they will be treated separately.
Kind regards

sbutcher-arm commented:
@adustm - Thanks. The PR is fine by me now. There are outstanding changes but we should revisit these when we've done the necessary changes to mbed TLS.

adustm commented:
Hello @0xc0170 I think the 'need review' can be transformed into 'ready to merge'... :)

RonEld commented:
Hi @adustm 
I believe your question on the thread safety should be referred to @andresag01

adustm commented:
Hello @andresag01, @sbutcher-arm @RonEld 
About the comment from Andres

> There have been some improvements, but I do not think that this code is thread-safe. Normally, we achieve this in mbed TLS by using the threading API (threading.h). Please note that there is a PR to add support for this: #3150 

I have read PR#3150 but it is not clear to me what you are waiting for: shall I wait until this _thread_ PR is integrated in mbed-tls then in mbed-os ? What would be the impact to my code ?

Or can it be treated separately in another PR later on ? I have work about AES for 2 other platforms + SHA1 and SHA256 that are pending the integration of the current PR. I would like to be able to progress on the subject.
I think 0xc0170 is waiting for your green light.
Kind regards

sbutcher-arm commented:
As I said above, @andresag01's points are valid, but I don't think we should hold up this PR for the work on thread safety it depends on. 

@0xc0170 - this PR is ready to merge. We know it has issues, but they're dependent on changes in mbed TLS and can be fixed later.

0xc0170 commented:
/morph test-nighly

0xc0170 commented:
/morph test-nightly

screamerbg commented:
/morph test

0xc0170 commented:
@bridadan Need you eye. The failures are about  sections overflow. 

@adustm The selftest change should not affect other targets or? You can view the failures : http://mbedosci.cloudapp.net/results/pr/3691/1660

From the failures, looks like the test does not fit to many devices that are tested.

adustm commented:
Hello,
I can see that some failure are due to code size. I've added AES test in the self test to show how I have validated the modification, but I can remove this file from the PR.

adustm commented:
Thanks @0xc0170 for sharing the build output failure. 
I have fixed it. It should be ok now.
Cheers

0xc0170 commented:
/morph test-nightly
